### PR TITLE
Enhancements and new Features

### PR DESCRIPTION
**Pull request by aristofanischionis**
A list of changes including: 
* Use percentages in positioning
* Clone slides
* Add new slide adds two text boxes
* Delete items completely restructured
* Position correctly where new textboxes and images spawn 
* Make the border of the text box visible around the real area
* Delete Images only when they are not referenced by any slide anymore

### CODE REVIEW

File: packages/slides-frontend/src/containers/Editor/components/image/Upload/Dropzone.js [hunk 1]
```diff
@@ -117,7 +117,7 @@ Dropzone.propTypes = {
 function mapDispatchToProps(dispatch) {
   return {
     onImageRequest: () => dispatch(uploadImageRequest(false)),
-    onAddImage: src => dispatch(addItem({type: ItemTypes.IMAGE, src})),
+    onAddImage: Src => dispatch(addItem({type: ItemTypes.IMAGE, Src})),
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/Editor/components/image/Upload/Dropzone.js:120]: Why this change? Does it really have to be uppercased?

File: packages/slides-frontend/src/containers/Editor/components/resize_move/index.css [hunk 2]
```diff
@@ -0,0 +1,9 @@
+.item-style:hover {
+    border-color: #66afe9;
+    box-shadow: 0 0 8px rgb(0, 140, 255);
+  }
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/Editor/components/resize_move/index.css]: wrong indentation

File: packages/slides-frontend/src/containers/Editor/components/resize_move/index.css [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+.item-style:focus {
+    border-color: #66afe9;
+    box-shadow: 0 0 8px rgb(0, 140, 255);
+  }
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/Editor/components/resize_move/index.css]: wrong indentation

File: packages/slides-frontend/src/containers/Editor/components/resize_move/index.js [hunk 4]
```diff
@@ -2,19 +2,27 @@ import React, {useState} from 'react';
 import PropTypes from 'prop-types';
 import reactable from 'reactablejs';
 import interact from 'interactjs';
-
 import {connect} from 'react-redux';
-
-import MyItem from '../item';
+import { Label }  from 'semantic-ui-react'
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/Editor/components/resize_move/index.js]:
```suggestion
import {Label}  from 'semantic-ui-react'
```

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/definitions.ts [hunk 5]
```diff
@@ -36,18 +34,20 @@ export class Text extends baseItem {
       this.type = ItemTypes.TEXT
       this.Data = "<p></p>\n";
       this.Edit = false;
-      this.Size = { width: 500, height: 80 };
+      this.Position= {x: 0.30, y: 0.40};
+      this.Size = { width: 0.25, height: 0.05 };
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/redux-store/DeckReducer/definitions.ts:38]: You are using `{ x }` and `{x}`. You should just stick to one of them :P

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/definitions.ts [hunk 6]
```diff
@@ -18,12 +18,10 @@ class baseItem {
     ID: string
     Position: position
     Size: size
-    Focused: boolean
     constructor(){
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/redux-store/DeckReducer/definitions.ts]:
```suggestion
    constructor() {
```

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/definitions.ts [hunk 7]
```diff
@@ -36,18 +34,20 @@ export class Text extends baseItem {
...
+      this.Position= {x: 0.30, y: 0.40};
+      this.Size = { width: 0.25, height: 0.05 };
     }
   }
 
 export class Image extends baseItem {
     readonly type: string
     Src: string
-    constructor(src: string){
+    constructor(Src: string){
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/redux-store/DeckReducer/definitions.ts]:
```suggestion
    constructor(Src: string) {
```

File: packages/slides-frontend/src/utils/helperFunctions.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+// check if the image name string appears in state
+// if only one, it means that it is the last time the image appears in the slides and has to be deleted
+
+const deleteImageFromBackend = (imageName, slidesStringified) => (slidesStringified.match(new RegExp(imageName,"g")) || []).length === 1;
```
Comment (mic4ael) [inline packages/slides-frontend/src/utils/helperFunctions.js]:
```suggestion
const deleteImageFromBackend = (imageName, slidesStringified) => (slidesStringified.match(new RegExp(imageName, "g")) || []).length === 1;
```

Comment (mic4ael) [reply packages/slides-frontend/src/utils/helperFunctions.js]: Hm.. that might not be the best way of checking if there are more than 1 images with the same name. What if e.g. I add the text element with the image name as the content.

Comment (mic4ael) [inline packages/slides-frontend/src/utils/helperFunctions.js]: I would rename it to `canDeleteImageFromBackend`

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts [hunk 9]
```diff
@@ -37,6 +37,43 @@ const DeckState = (state: Deck = initialDeck, action: Action): Deck =>
           ID: uuidv4(),
           itemsArray: [], // add a box that can be image or text
         }
+
+        // this code is to add two text Boxes by default in any new slide
+        // START
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: That could be extracted to a separate method/function.

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts [hunk 10]
```diff
@@ -37,6 +37,43 @@ const DeckState = (state: Deck = initialDeck, action: Action): Deck =>
...
+        itm2.Position = {
+          x: 0.225,
+          y: 0.30,
+        }
+        itm2.Size = {
+          width: 0.40,
+          height: 0.40,
+        }
+        slide.itemsArray.push(itm1);
+        slide.itemsArray.push(itm2);
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: you can just do it in one call to push - `slide.itemsArray.push(itm1, itm2)`

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts [hunk 11]
```diff
@@ -47,6 +84,37 @@ const DeckState = (state: Deck = initialDeck, action: Action): Deck =>
...
+      case CLONE_SLIDE: {
+        // create a new blank slide
+        let slide:Slide = {
+          ID: uuidv4(),
+          itemsArray: [],
+        }
+        // with a loop over the current itemsArray, create and push newItems in the new itemsArray
+        slide.itemsArray = draft.slides[draft.currentSlide].itemsArray.map(currentItem => {
+          const copyObj = JSON.parse(JSON.stringify(currentItem));
+          if(copyObj.type === ItemTypes.TEXT) {
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]:
```suggestion
          if (copyObj.type === ItemTypes.TEXT) {
```

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts [hunk 12]
```diff
@@ -47,6 +84,37 @@ const DeckState = (state: Deck = initialDeck, action: Action): Deck =>
...
       }
+      case CLONE_SLIDE: {
+        // create a new blank slide
+        let slide:Slide = {
+          ID: uuidv4(),
+          itemsArray: [],
+        }
+        // with a loop over the current itemsArray, create and push newItems in the new itemsArray
+        slide.itemsArray = draft.slides[draft.currentSlide].itemsArray.map(currentItem => {
+          const copyObj = JSON.parse(JSON.stringify(currentItem));
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: if currentItem is an object you could do something like `const copyObj = {...currentItem}` to copy an object. But keep in mind this will perform a shallow copy

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts [hunk 13]
```diff
@@ -47,6 +84,37 @@ const DeckState = (state: Deck = initialDeck, action: Action): Deck =>
...
+          if(copyObj.type === ItemTypes.TEXT) {
+            const itemObj: Text = new Text();
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: It would be nice if `Text` had a `clone` method.

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts [hunk 14]
```diff
@@ -47,6 +84,37 @@ const DeckState = (state: Deck = initialDeck, action: Action): Deck =>
...
+            itemObj.Data = copyObj.Data;
+            // change size and position
+            itemObj.Size = copyObj.Size;
+            itemObj.Position = copyObj.Position;
+            return itemObj;
+          }
+
+          // if it is IMAGE
+          const { Src } = copyObj;
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: `clone` method on those classes would make it much more generic and extensible.

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts (hunk 12 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: I need a deep copy, that's why I do it like that.
The `currentItem` object has Size and Position sub-objects that will not be identified as objects if I do it another way.
They will remain `Proxies`

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts (hunk 13 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: In the definitions I have a `newItem()` method, that gets an object and tries to create the appropriate `Text` or `Image` type.
I could use this instead, but I had some issues because I use TypeScript.
it would be nice to do:
`const itemObj: Text = newItem({type: "TEXT"})`
but TypeScript complains, because the newItem returns the mother class `Item`.
I need to have it as Text object in the reducer at the CLONE_SLIDE case in order to access its `Data` attribute.

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts (hunk 12 above)
Comment (mic4ael) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: Yeah, I am aware of it. It is just that it looks a bit ugly IMO :P You could write a custom function (which wouldn't be hard to do) that would perform a deep clone on an object.

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts (hunk 14 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: Same as above.
Is the best approach to create the two clone methods?

Comment (mic4ael) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: the best would be to have an interface/base class of all those classes .e.g. `Clonable` that would define a single method `clone`. Then it would be up to a specific inheriting class how this method would be implemented

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts (hunk 10 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: :+1:

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts (hunk 11 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: :+1:

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts (hunk 9 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: :+1:

File: packages/slides-frontend/src/utils/helperFunctions.js (hunk 8 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/utils/helperFunctions.js]: :+1:

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/definitions.ts (hunk 6 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/definitions.ts]: :+1:

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/definitions.ts (hunk 5 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/definitions.ts:38]: :+1:

File: packages/slides-frontend/src/containers/Editor/components/resize_move/index.js (hunk 4 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/Editor/components/resize_move/index.js]: :+1:

File: packages/slides-frontend/src/containers/Editor/components/resize_move/index.css (hunk 2 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/Editor/components/resize_move/index.css]: :+1:

File: packages/slides-frontend/src/containers/Editor/components/resize_move/index.css (hunk 3 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/Editor/components/resize_move/index.css]: :+1:

File: packages/slides-frontend/src/containers/Editor/components/image/Upload/Dropzone.js (hunk 1 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/Editor/components/image/Upload/Dropzone.js:120]: Yes, because in the reducer of the DeckReducer  I can get the `Src` attribute of the object and assign it directly to the Image object. Now everywhere that the src of the image is referenced the `Src` keyword is used.

Comment (mic4ael) [reply packages/slides-frontend/src/containers/Editor/components/image/Upload/Dropzone.js:120]: Yeah, I saw it. I am just not sure why this even has to be uppercase in the class where it is used :P

File: packages/slides-frontend/src/utils/helperFunctions.js (hunk 8 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/utils/helperFunctions.js]: Well, the text is saved as html in the redux store, in order to keep the styles.
So even if there are no styles, the text will have `<p>` tags around it.
A workaround would be, instead of checking for the `imageName` string, to search for the  `"type":"IMAGE","Src":"${imageName}"` string. This way even if the user has written the same text in a text box, in the stringified state it will be like: 
```
"type":"TEXT","Data":"<p>\"type\":\"IMAGE\",\"Src\":\"5ae3c763551cbea7ec39302742ef8455_me.JPG\"</p>\n"
```
which doesn't match with the image name, so it doesn't get confused.
I tried it and works, is it good enough?

Comment (mic4ael) [reply packages/slides-frontend/src/utils/helperFunctions.js]: Don't you think it is a bit overcomplicated? Why don't you just iterate over all images elements and check if their name property is the same?

File: packages/slides-frontend/src/containers/Editor/components/image/Upload/Dropzone.js (hunk 1 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/Editor/components/image/Upload/Dropzone.js:120]: hm more about consistency I guess, should I change it everywhere?
Also the Position, Size, Data, Edit, ID ?

Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/Editor/components/image/Upload/Dropzone.js:120]: but okay I can change only the Src if you want

File: packages/slides-frontend/src/utils/helperFunctions.js (hunk 8 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/utils/helperFunctions.js]: ok so get the state as an object and have a loop?
I think it will be more complicated but okay I can do it no problem

File: packages/slides-frontend/src/containers/Editor/components/image/Upload/Dropzone.js (hunk 1 above)
Comment (mic4ael) [reply packages/slides-frontend/src/containers/Editor/components/image/Upload/Dropzone.js:120]: no, no, you don't have to change it :) I was more curious why it started like that ;) But let's keep it as it is.

Comment (mic4ael) [reply packages/slides-frontend/src/containers/Editor/components/image/Upload/Dropzone.js:120]: And what I meant is that you could have something like

```javascript
onAddImage: src => dispatch(addItem({type: ItemTypes.IMAGE, Src: src})),
```

File: packages/slides-frontend/src/utils/helperFunctions.js (hunk 8 above)
Comment (mic4ael) [reply packages/slides-frontend/src/utils/helperFunctions.js]: I think it will be better in the long run and much easier to understand in the future.

File: packages/slides-frontend/src/containers/Editor/components/image/Upload/Dropzone.js (hunk 1 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/Editor/components/image/Upload/Dropzone.js:120]: ah yes okay! :smile: 
So I leave it as is for now.

File: packages/slides-frontend/src/utils/helperFunctions.js (hunk 8 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/utils/helperFunctions.js:4]:
``` js
const imageArray = slides.map(slide =>
        slide.itemsArray.map(item => {
            if (item.type === "IMAGE" && item.Src === imageName) {
                return item.Src;
            }
        })
    );
```

this doesn't give me the expected results... Any idea?

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts (hunk 12 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: Ok fixed it

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts (hunk 14 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts:106]: So the order of inheritance should be:
* Clonable
    * baseItem
        * Text
        * Image
and the clone method: 
``` js
public clone():Clonable {
    let copy = new (this.constructor());
    copy = deepCopyFunction(this)
    return copy;
  }
```

and then the itemsArray contains Items, which is
`export type Item = Text | Image;`
but I want it to get Clonable objects instead.

Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: Actually we don't need cloning and extra stuff, we can just deepCopy the Item and push it in the itemsArray of the new Slide.
```js
slide.itemsArray = draft.slides[draft.currentSlide].itemsArray.map(currentItem => {
  const copyObj: Item = deepCopyFunction(currentItem);
  return copyObj;
});
```

File: packages/slides-frontend/src/utils/helperFunctions.js (hunk 8 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/utils/helperFunctions.js]: ok did it

Comment (aristofanischionis) [reply packages/slides-frontend/src/utils/helperFunctions.js]: we have to consider what happens in the remove slide, when the pictures should be deleted from backend.

Comment (aristofanischionis) [reply packages/slides-frontend/src/utils/helperFunctions.js]: Right now, if a slide has a picture and is the last one that has it, the picture file doesn't get deleted in the backend.

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts (hunk 14 above)
Comment (aristofanischionis) [reply packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]: I did it like that and seems to be working.

**Comment by aristofanischionis** [comment]
@mic4ael I think now the code is in a good condition. Two things that still annoy me are:
* The resize and move still has some glitches and is not very accurate especially when resizing.
* The resize-move/index.js file is large and should be split into different functions and files IMO.

File: packages/slides-frontend/public/index.html [hunk 15]
```diff
@@ -23,7 +23,7 @@
       work correctly both with client-side routing and a non-root public URL.
       Learn how to configure a non-root public URL by running `npm run build`.
     -->
-    <title>Slides App</title>
+    <title>Slides' App</title>
```
Comment (mic4ael) [inline packages/slides-frontend/public/index.html]: I don't think this is correct.

File: packages/slides-frontend/src/utils/helperFunctions.js [hunk 16]
```diff
@@ -0,0 +1,33 @@
+
+const canDeleteImageFromBackend = (imageName, slidesStringified) => {
+    const slides = JSON.parse(slidesStringified);
```
Comment (mic4ael) [inline packages/slides-frontend/src/utils/helperFunctions.js]: Why don't you pass `slides` array of objects directly? There is no point now in using `slidesStringified`

File: packages/slides-frontend/src/utils/helperFunctions.js [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    let counter = 0;
+    slides.map(slide =>
```
Comment (mic4ael) [inline packages/slides-frontend/src/utils/helperFunctions.js]: use `slides.forEach` since `slides.map` is supposed to return something.

File: packages/slides-frontend/src/utils/helperFunctions.js [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+            }
+        })
+    );
+    return counter === 1;
+}
+
+const deepCopyFunction = inObject => {
+    let outObject, value, key;
+    if (typeof inObject !== "object" || inObject === null) {
+      return inObject; // Return the value if inObject is not an object
```
Comment (mic4ael) [inline packages/slides-frontend/src/utils/helperFunctions.js]: I think some of these comments are not really necessary. The code is pretty self-explanatory.

File: packages/slides-frontend/src/containers/Editor/SideBar/index.js [hunk 19]
```diff
@@ -3,27 +3,58 @@ import {connect} from 'react-redux';
...
+
   const addingSlide = () => {
     onAddSlide();
-    window.location = `#/${currentSlide + 1}`; // because slides here are starting from 1
+    window.location = `#/${currentSlide + 1}`;
   };
+
   const removingSlide = () => {
+    // check if pictures should be deleted from backend
+    itemsArray.map(item => {
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/Editor/SideBar/index.js]: `itemsArray.forEach` - `map` is used when you want to return a result of the iteration

File: packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts [hunk 20]
```diff
@@ -47,6 +83,20 @@ const DeckState = (state: Deck = initialDeck, action: Action): Deck =>
...
       }
+      case CLONE_SLIDE: {
+        // create a new blank slide
+        let slide:Slide = {
+          ID: uuidv4(),
+          itemsArray: [],
+        }
+        slide.itemsArray = draft.slides[draft.currentSlide].itemsArray.map(currentItem => {
+          const copyObj: Item = deepCopyFunction(currentItem);
+          return copyObj;
```
Comment (mic4ael) [inline packages/slides-frontend/src/containers/redux-store/DeckReducer/reducer.ts]:
```suggestion
          return deepCopyFunction(currentItem);
```
